packages/cdk/resources/CognitoFunctions.ts

```diff
@@ -0,0 +1,237 @@
+  readonly mockOidcUserInfoEndpoint?: string
+  readonly mockOidcjwksEndpoint?: string
+  readonly mockOidcClientId?: string
+  readonly mockOidcIssuer?: string
+  readonly tokenMappingTable: ITableV2
+  readonly tokenMappingTableWritePolicy: IManagedPolicy
+  readonly tokenMappingTableReadPolicy: IManagedPolicy
+  readonly useTokensMappingKmsKeyPolicy: IManagedPolicy
+  readonly primaryPoolIdentityProviderName: string
+  readonly mockPoolIdentityProviderName: string
+  readonly logRetentionInDays: number
+  readonly deploymentRole: IRole
+
+}
+
+/**
+ * Functions and resources that are needed for cognito
+ */
+export class CognitoFunctions extends Construct {
+  public readonly cognitoPolicies: Array<IManagedPolicy>
+  public readonly tokenLambda: NodejsFunction
+  public readonly mockTokenLambda: NodejsFunction
+  public readonly primaryJwtPrivateKey: Secret
+
+  public constructor(scope: Construct, id: string, props: CognitoFunctionsProps) {
+    super(scope, id)
+
+    // Resources
+
+    // kms key used to encrypt the secret that stores the JWT private key
+    const jwtKmsKey = new Key(this, "JwtKMSKey", {
+      removalPolicy: RemovalPolicy.DESTROY,
+      pendingWindow: Duration.days(7),
+      alias: `${props.stackName}-JwtKMSKeyKMSKey`,
+      description: `${props.stackName}-JwtKMSKeyKMSKey`,
+      enableKeyRotation: true,
+      policy: new PolicyDocument({
+        statements: [
+          new PolicyStatement({
+            sid: "Enable IAM User Permissions",
+            effect: Effect.ALLOW,
+            actions: [
+              "kms:*"
+            ],
+            principals: [
+              new AccountRootPrincipal
+            ],
+            resources: ["*"]
+          }),
+          new PolicyStatement({
+            effect: Effect.ALLOW,
+            principals: [props.deploymentRole],
+            actions: [
+              "kms:Encrypt",
+              "kms:GenerateDataKey*"
+            ],
+            resources:["*"]
+          })
+        ]
+      })
+    })
+    jwtKmsKey.addAlias(`alias/${props.stackName}-jwtKmsKey`)
+    const useJwtKmsKeyPolicy = new ManagedPolicy(this, "UseJwtKmsKeyPolicy", {
+      statements: [
+        new PolicyStatement({
+          actions: [
+            "kms:DescribeKey",
+            "kms:Decrypt"
+          ],
+          resources: [
+            jwtKmsKey.keyArn
+          ]
+        })
+      ]
+    })
+
+    // define some variables that we need if we are doing mock authorization
+    let mockTokenLambda: LambdaFunction
+
+    // set up things for mock authorization
+    if (props.useMockOidc) {
+      if (props.mockOidcjwksEndpoint === undefined ||
+        props.mockOidcUserInfoEndpoint === undefined ||
+        props.mockOidcTokenEndpoint === undefined ||
+        props.mockOidcClientId === undefined ||
+        props.mockOidcIssuer === undefined
+      ) {
+        throw new Error("Attempt to use mock oidc but variables are not defined")
+      }
```

Comment:
as above

Reply:
this is needed for typing.
The props in the if statement are optional to this construct and are only passed in when useMockOidc is true.
If we dont have the if statement, then typescript thinks they may be undefined and so throws errors when compiling the resources inside the if statement